chain/chain/src/flat_storage_resharder.rs

```diff
@@ -460,6 +476,180 @@ impl FlatStorageResharder {
+                resharder: self.clone(),
+                shard_uid,
+                flat_head_block_hash,
+            });
+            return FlatStorageReshardingTaskStatus::Cancelled;
+        }
+
+        // If the flat head is not in the canonical chain this task has failed.
+        match chain_store.get_block_hash_by_height(height) {
+            Ok(hash) => {
+                if hash != flat_head_block_hash {
+                    error!(target: "resharding", ?shard_uid, ?height, ?flat_head_block_hash, ?hash, "flat head not in canonical chain!");
+                    return FlatStorageReshardingTaskStatus::Failed;
+                }
+            }
+            Err(err) => {
+                error!(target: "resharding", ?shard_uid, ?err, ?height, "can't find block hash by height");
+                return FlatStorageReshardingTaskStatus::Failed;
+            }
+        }
+
+        // Apply deltas and then create the flat storage.
+        let apply_result =
+            self.shard_catchup_apply_deltas(shard_uid, flat_head_block_hash, chain_store);
+        let Ok((num_batches_done, flat_head)) = apply_result else {
+            error!(target: "resharding", ?shard_uid, err = ?apply_result.unwrap_err(), "flat storage shard catchup delta application failed!");
+            return FlatStorageReshardingTaskStatus::Failed;
+        };
+        match self.shard_catchup_finalize_storage(shard_uid, &flat_head) {
+            Ok(_) => {
+                let task_status = FlatStorageReshardingTaskStatus::Successful { num_batches_done };
+                info!(target: "resharding", ?shard_uid, ?task_status, "flat storage shard catchup task finished");
+                // At this point we can trigger the reload of memtries.
+                self.scheduler.send(ReshardingRequest::MemtrieReload { shard_uid });
+                task_status
+            }
+            Err(err) => {
+                error!(target: "resharding", ?shard_uid, ?err, "flat storage shard catchup finalize failed!");
+                FlatStorageReshardingTaskStatus::Failed
+            }
+        }
+    }
+
+    /// Applies flat storage deltas in batches on a shard that is in catchup status.
+    ///
+    /// Returns the number of delta batches applied and the final tip of the flat storage.
+    fn shard_catchup_apply_deltas(
+        &self,
+        shard_uid: ShardUId,
+        mut flat_head_block_hash: CryptoHash,
+        chain_store: &ChainStore,
+    ) -> Result<(usize, Tip), Error> {
+        const CATCH_UP_BLOCKS: u32 = 50;
+
+        let mut num_batches_done: usize = 0;
+
+        loop {
+            let _span = tracing::debug_span!(
+                target: "resharding",
+                "shard_catchup_apply_deltas/batch",
+                ?shard_uid,
+                ?flat_head_block_hash,
+                batch_id = ?num_batches_done)
+            .entered();
+            let chain_final_head = chain_store.final_head()?;
```

Comment:
Cool that you refresh this from within the loop, it will get as close to final as possible.